File: jal-rakshak-frontend/src/api/telemetry.ts
import { apiFetch } from "./client";

export interface PointAData {
  tds: number;
  turbidity: number;
  temperature: number;
}

export interface PointBData {
  ph: number;
  tds: number;
  turbidity: number;
  temperature: number;
}

export interface LatestTelemetry {
  id: number;
  device_id: string;
  timestamp: string;

  point_a: PointAData;

  point_b: PointBData;

  flow_rate_l_min: number;
  pump_status: boolean;
}

export interface LatestTelemetryResponse {
  success: boolean;
  data: LatestTelemetry;
}

export async function getLatestTelemetry(
  deviceId?: string
): Promise<LatestTelemetryResponse> {
  const endpoint = deviceId
    ? `/api/telemetry/latest?device_id=${encodeURIComponent(deviceId)}`
    : '/api/telemetry/latest';

  return apiFetch<LatestTelemetryResponse>(endpoint);
}

File: jal-rakshak-backend/app/schemas/telemetry.py
from datetime import datetime

from pydantic import BaseModel, Field


class PointAReading(BaseModel):
    tds: float = Field(..., ge=0)
    turbidity: float = Field(..., ge=0)
    temperature: float


class PointBReading(BaseModel):
    ph: float = Field(..., ge=0, le=14)
    tds: float = Field(..., ge=0)
    turbidity: float = Field(..., ge=0)
    temperature: float


class TelemetryPayload(BaseModel):
    device_id: str = Field(..., min_length=1)
    timestamp: datetime

    point_a: PointAReading
    point_b: PointBReading

    flow_rate: float = Field(..., ge=0, le=2)

    pump_status: bool